api/src/wfl/api/handlers.clj

```diff
@@ -118,65 +112,72 @@
 (defn post-create
   "Create the workload described in BODY of REQUEST."
   [{:keys [parameters] :as request}]
-  (wfl.api.handlers/print-handler-error
-   (letfn [(unnilify [m] (into {} (filter second m)))]
-     (let [{:keys [body]} parameters
-           {:keys [email]} (gcs/userinfo request)]
-       (logr/infof "post-create endpoint called: body=%s" body)
-       (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
-         (->> (assoc body :creator email)
-              (add-workload! tx)
-              :uuid
-              (conj ["SELECT * FROM workload WHERE uuid = ?"])
-              (jdbc/query tx)
-              first unnilify succeed))))))
+  (log-and-fail-on-error
+    (let [{:keys [body]} parameters
+          {:keys [email]} (gcs/userinfo request)]
+      (logr/infof "post-create endpoint called: body=%s" body)
+      (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+        (->>
+          (assoc body :creator email)
+          (add-workload! tx)
+          :uuid
+          (postgres/load-workload-for-uuid tx)
+          succeed)))))
 
 (defn get-workload!
   "List all workloads or the workload with UUID in REQUEST."
   [request]
-  (wfl.api.handlers/print-handler-error
-   (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
-     (let [uuid (get-in request [:parameters :query :uuid])
-           uuid-vec (->> (if uuid
-                         (do
-                           (logr/infof "get-workload endpoint called: uuid=%s" uuid)
-                           [{:uuid uuid}])
-                         (do
-                           (logr/info "get-workload endpoint called with no uuid, querying all")
-                           (jdbc/query tx ["SELECT uuid FROM workload"]))))]
-       (if (not-empty uuid-vec)
-         (do
-           (util/do-or-nil (postgres/update-workload-for-uuid! tx uuid-vec))
-           (let [result (mapv (partial postgres/get-workload-for-uuid tx) uuid-vec)]
-             (if (every? empty? result)
-               (fail-with-response response/not-found (format "Workload %s not found" uuid))
-               (succeed result))))
-         (success []))))))
+  (letfn [(go! [tx {:keys [uuid id] :as workload}]
+            (if-not (:finished workload)
+              (do
+                (logr/infof "updating workload %s" uuid)
+                (postgres/update-workload! tx workload)
+                (postgres/load-workload-for-id tx id))
+              workload))]
+    (log-and-fail-on-error
+      (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+        (succeed
+          (mapv (partial go! tx)
+            (if-let [uuid (-> request :parameters :query :uuid)]
+              (if-let [workload (postgres/load-workload-for-uuid tx uuid)]
```

Comment:
@tbl3rd any suggestions for combining the nested if?

Reply:
Given they have distinct alternates, nested `if-let` seems the way to go.

Reply:
Why `mapv` instead of `map`?

Reply:
Because i need to force evaluation before the transaction exits